Use checksums instead of fsyncs to avoid slow discovery caching on MacOS

negz opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind bug

#### What this PR does / why we need it:

Short version: Avoids kubectl "hanging" for 10+ seconds while writing to cache when an API server has many API groups.

Part of the API discovery cache uses an HTTP RoundTripper that transparently caches responses to disk. The upstream implementation of the disk cache is hard coded to call `Sync()` on every file it writes. This has noticeably poor performance on modern Macs, which ask their  controllers to flush all the way to persistent storage because Go uses the `F_FULLFSYNC` fnctl. Apple recommends minimizing this behavior in order to avoid degrading performance and increasing disk wear.

The content of the discovery cache is not critical; it is indeed just a cache and can be recreated by hitting the API servers' discovery endpoints. This commit replaces upstream httpcache's diskcache implementation with a similar implementation that can use SHA256 sums to detect corrupted cache entries at read-time. When such an entry is detected (e.g. because it was only partially flushed to permanent storage before the host lost power) the cache will report a miss. This causes httpcache to fall back to its underlying HTTP transport (i.e. the real API server) and re-cache the resulting value.

Apart from adding SHA256 sums and avoiding calling fsync this implementation differs from upstream httpcache's diskcache package in that it also uses SHA256 sums rather than MD5 sums of cache keys in order to sanitize cache filenames.

Note that I'm adding this implementation here rather than submitting it upstream because httpcache does not appear to have been touched for the last three years - it seems unlikely that they are accepting contributions.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubernetes/issues/110753

#### Special notes for your reviewer:

Since this is intended to improve slow performance I've added a benchmark test. You can see there's a dramatic speed improvement on MacOS:

```console
# Before
$ go test -v -bench . -run '^Bench'
goos: darwin
goarch: arm64
pkg: k8s.io/client-go/discovery/cached/disk
BenchmarkDiskCache
BenchmarkDiskCache-10    	      60	  22272642 ns/op
PASS
ok  	k8s.io/client-go/discovery/cached/disk	2.582s

# After
$ go test -v -bench . -run '^Bench'
goos: darwin
goarch: arm64
pkg: k8s.io/client-go/discovery/cached/disk
BenchmarkDiskCache
BenchmarkDiskCache-10    	    1534	    761469 ns/op
PASS
ok  	k8s.io/client-go/discovery/cached/disk	1.483s
```

I also see a decent speed improvement on Linux, though note that I'm running this test inside a Parallels VM:

```console
# Before
$ go test -v -bench . -run '^Bench'
goos: linux
goarch: arm64
pkg: k8s.io/client-go/discovery/cached/disk
BenchmarkDiskCache
BenchmarkDiskCache-6         812           1316323 ns/op
PASS
ok      k8s.io/client-go/discovery/cached/disk  1.240s

# After
$ go test -v -bench . -run '^Bench'
goos: linux
goarch: arm64
pkg: k8s.io/client-go/discovery/cached/disk
BenchmarkDiskCache
BenchmarkDiskCache-6        1513            772180 ns/op
PASS
ok      k8s.io/client-go/discovery/cached/disk  1.267s
```

Finally, invoking kubectl against an API server with 300+ API groups now takes about a second rather than 10+ (as demonstrated in https://github.com/kubernetes/kubernetes/issues/110753).

```console
$ bin/kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"25+", GitVersion:"v1.25.0-alpha.2.14+4fddaf7eb2cf4e", GitCommit:"4fddaf7eb2cf4ee904e500f4af4f01f0a409d19f", GitTreeState:"clean", BuildDate:"1980-01-01T00:00:00Z", GoVersion:"go1.18.2", Compiler:"gc", Platform:"darwin/arm64"}
Kustomize Version: v4.5.4
Server Version: version.Info{Major:"1", Minor:"22+", GitVersion:"v1.22.9-eks-a64ea69", GitCommit:"540410f9a2e24b7a2a870ebfacb3212744b5f878", GitTreeState:"clean", BuildDate:"2022-05-12T19:15:31Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}
WARNING: version difference between client (1.25) and server (1.22) exceeds the supported minor version skew of +/-1

$ rmd ~/.kube/cache

# Cold cache
$ time bin/kubectl get nodes
NAME                                            STATUS   ROLES    AGE    VERSION
ip-192-168-103-188.us-west-2.compute.internal   Ready    <none>   7d6h   v1.22.9-eks-810597c
ip-192-168-63-15.us-west-2.compute.internal     Ready    <none>   7d6h   v1.22.9-eks-810597c
ip-192-168-93-0.us-west-2.compute.internal      Ready    <none>   7d6h   v1.22.9-eks-810597c
bin/kubectl get nodes  0.41s user 0.47s system 61% cpu 1.428 total

# Warm cache
$ time bin/kubectl get nodes
NAME                                            STATUS   ROLES    AGE    VERSION
ip-192-168-103-188.us-west-2.compute.internal   Ready    <none>   7d6h   v1.22.9-eks-810597c
ip-192-168-63-15.us-west-2.compute.internal     Ready    <none>   7d6h   v1.22.9-eks-810597c
ip-192-168-93-0.us-west-2.compute.internal      Ready    <none>   7d6h   v1.22.9-eks-810597c
bin/kubectl get nodes  0.17s user 0.05s system 160% cpu 0.139 total

$ bin/kubectl api-versions|wc -l
     343

$ du -hd1 ~/.kube/cache
1.7M    /Users/negz/.kube/cache/discovery
1.7M    /Users/negz/.kube/cache/http
3.4M    /Users/negz/.kube/cache
```

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Use checksums instead of fsyncs to ensure discovery cache integrity
```

negz committed:
Use checksums instead of fsyncs to manage discovery cache corruption

Part of the API discovery cache uses an HTTP RoundTripper that
transparently caches responses to disk. The upstream implementation of
the disk cache is hard coded to call Sync() on every file it writes.
This has noticably poor performance on modern Macs, which ask their disk
controllers to flush all the way to persistant storage because Go uses
the `F_FULLFSYNC` fnctl. Apple recommends minimizing this behaviour in
order to avoid degrading performance and increasing disk wear.

The content of the discovery cache is not critical; it is indeed just a
cache and can be recreated by hitting the API servers' discovery
endpoints. This commit replaces upstream httpcache's diskcache
implementation with a similar implementation that can use CRC-32
checksums to detect corrupted cache entries at read-time. When such an
entry is detected (e.g. because it was only partially flushed to
permanent storage before the host lost power) the cache will report a
miss. This causes httpcache to fall back to its underlying HTTP
transport (i.e. the real API server) and re-cache the resulting value.

Apart from adding CRC-32 checksums and avoiding calling fsync this
implementation differs from upstream httpcache's diskcache package in
that it uses FNV-32a hashes rather than MD5 hashes of cache keys in
order to generate filenames.

Signed-off-by: Nic Cope <nicc@rk0n.org>

negz committed:
Add a benchmark for the discovery cache RoundTripper

This benchmark is intended to demonstrate a performance improvement
gained by removing fsyncs. Refer to the below issue for more detail.

https://github.com/kubernetes/kubernetes/issues/110753

Signed-off-by: Nic Cope <nicc@rk0n.org>

leilajal commented:
/assign @Jefftree 
/triage accepted

Jefftree commented:
/ok-to-test

Jefftree commented:
/cc @lavalamp @seans3

lavalamp commented on staging/src/k8s.io/client-go/discovery/cached/disk/round_tripper.go:
```diff
@@ -63,3 +66,54 @@ func (rt *cacheRoundTripper) CancelRequest(req *http.Request) {
+	h := fnv.New32a()
+	_, _ = h.Write([]byte(key)) // Writing to a hash never returns an error.
+	return fmt.Sprintf("%X", h.Sum32())
+}
```
The reason to move away from MD5 is that it's broken. I think there is not really a reason to avoid cryptographic hashes? The worst thing that happens is maybe you accidentally prevent an attack that none of us have thought of?

negz replied:
I don't feel strongly here, but using a cryptographic hash for non-sensitive data just in case feels a bit odd to me. The reason we hash here is simply to turn a request method and URL into a valid filename. The content of the file isn't hashed, so using a cryptographic hash would only serve to obfuscate what URL the content came from at the expense of a bit of speed.

mpuckett159 replied:
Yeah not sure I understand why we would want to use a cryptographic hash algo here. Maybe lower chance of collision but still seems like a bit much.

lavalamp replied:
Cryptographic hashes get as close as possible to a guarantee there's no bad values, or e.g. in this use, chunks of data that could be deleted without changing the sum. They are really not much more expensive than bad hash functions. Even though this cache is not that important, I think the extra computation is more than paid for by the extra guarantees.

Second point, I find "security mindset" is best understood as not creating *ingredients* for attacks. Non-cryptographic hashes are an attack ingredient. The remaining ingredients would perhaps be an unexpected use of this cache, and unexpected control of contents by attacker.

negz replied:
Pushed a commit that switches to sha256.

lavalamp commented on staging/src/k8s.io/client-go/discovery/cached/disk/round_tripper.go:
```diff
@@ -63,3 +66,54 @@ func (rt *cacheRoundTripper) CancelRequest(req *http.Request) {
+		return []byte{}, false
+	}
+
+	response := b[binary.MaxVarintLen32:]
```
Wouldn't it be simpler to just hash the bytes of the file and verify they match the key?

lavalamp replied:
(I'm not worried about performance because a cache miss means a network trip, which is going to be much slower)

negz replied:
@lavalamp I think you might have misinterpreted what's happening here. The first `binary.MaxVarintLen32` of the file contains a CRC-32 checksum of the (entire) remainder of the file, which is the actual cached content.

The key `httpcache` passes to this method is the HTTP method and URL of the request it wants to lookup in the cache (e.g. `GET https://apiserver.example.org:8080/apis/v1/coolstuff`). Per my response below the only reason we hash that key is to sanitize it to ensure it's a valid filename.

Would it help to update the Godoc string or add additional commentary to make this clearer?

negz replied:
@lavalamp FWIW that was my intuition too when I first looked into this. 😄

lavalamp commented:
Feedback:
1. Hash the whole file
2. Use a cryptographic hash (note that, if you don't like how big they are, it's OK to only use some of the bytes for the key)

mpuckett159 commented on staging/src/k8s.io/client-go/discovery/cached/disk/round_tripper.go:
```diff
@@ -63,3 +66,54 @@ func (rt *cacheRoundTripper) CancelRequest(req *http.Request) {
+}
+
+func (c *crcDiskCache) Delete(key string) {
+	_ = c.disk.Erase(sanitize(key)) // Nothing we can do with this error.
```
Did some quick testing on this to see what would happen if the file it was trying to Erase wasn't found and it just hands back a "file not found" error. We could possible just get the error back, check that it gives us a "file not found" and ignore since that doesn't matter, hand back the error if it isn't that one. I'm not really sure how much it matters though.

negz replied:
I imagine the best we could do if we got an error beside file not found would be to retry the delete. I don't feel strongly but I'm inclined to leave this as is - at the very least it's not a regression since this error has always been ignored:

https://github.com/gregjones/httpcache/blob/901d907/diskcache/diskcache.go#L38

mpuckett159 replied:
Yeah seems like it doesn't matter either way. If there's some other problem with access to the disk then the cache is the least of our worries.

negz committed:
Use sha256 to sanitize discovery HTTP cache keys

This helps avoid (potentially malicious) collisions when reading and
writing cache data.

Signed-off-by: Nic Cope <nicc@rk0n.org>

negz commented:
/test pull-kubernetes-e2e-kind

seans3 reviewed (approved):
/lgtm

**Major Caveat**: We need to add a release note. By changing the MD5 hash to the FNV-32a, we are stranding **all** currently cached OpenAPI and discovery documents. The current disk cache from Greg Jones uses MD5 to map the URL for a cached document to a unique filename. It looks like the MD5 hash is basically used as a `Base64 encode` substitute. So by changing this hash algorithm, all currently disk-cached files will no longer be found. OTOH, this is not as important as it sounds since both the OpenAPI (with cache-busting) and discovery (with aggregated discovery) document formats will soon be changed.

There is currently no code to clean up the disk cache, but this functionality will become increasingly important with these changes.

seans3 commented:
/priority important-soon

apelisse commented:
> So by changing this hash algorithm, all currently disk-cached files will no longer be found.

Yeah, we never use that for discovery (yet), and OpenAPI changes all the time. This is as bad as installing a CRD on a cluster. I don't know if it warrants a release note.

lavalamp commented on staging/src/k8s.io/client-go/discovery/cached/disk/round_tripper.go:
```diff
@@ -63,3 +66,55 @@ func (rt *cacheRoundTripper) CancelRequest(req *http.Request) {
+
+	response := b[binary.MaxVarintLen32:]
+	sum, _ := binary.Uvarint(b[:binary.MaxVarintLen32])
+	if crc32.ChecksumIEEE(response) != uint32(sum) {
```
I'd like to use SHA256 (or other strong hash) here, not just in the key. Sorry for not having been clear.

negz replied:
Ah got it. Will update.

lavalamp commented on staging/src/k8s.io/client-go/discovery/cached/disk/round_tripper.go:
```diff
@@ -63,3 +66,55 @@ func (rt *cacheRoundTripper) CancelRequest(req *http.Request) {
+	}
+
+	response := b[binary.MaxVarintLen32:]
+	sum, _ := binary.Uvarint(b[:binary.MaxVarintLen32])
```
BTW there's no reason to use variable lengths here, since any good hash will be a known size and will not be shorter than that most of the time.

negz committed:
Use SHA256 sums to verify discovery cache integrity

This is a little more computationally expensive but reduces the
likelihood of a potentially malicious cache collision.

Signed-off-by: Nic Cope <nicc@rk0n.org>

negz commented:
/retest

negz commented:
@lavalamp @seans3 PTAL - we're now SHA256 summing all the things. 😄

dims commented:
/retest

negz commented:
Hey folks, just checking in ahead of code freeze next week. Do you feel this is mergeable as is?

lavalamp commented:
/approve

sftim commented:
Should this PR have a release note?

csantanapr commented:
@negz Could you provide a one-liner for release notes?

negz commented:
@csantanapr Not sure how much I'd consider this a user facing change, but I've added a one-liner nonetheless. Thanks!

sftim commented:
Thanks @negz 

End users can observe the effects of this change without using debug tools.
